# ChEA-KG: Exploring a Human Gene Regulatory Network  
## **Table of Contents**
1. 1 - [Abstract](#abstract)
2. 2 - [The ChEA-KG GRN](#introduction)
3. 3 - [Searching a subnetwork the KG](#grn-search)
	1. &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1 - [Single Term Search](#single-search)
    2. &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2 - [Expanded Single Term Search](#subgraph-search)
	3. &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3 - [Two Term Search](#two-term-search)
4. 4 - [Interacting with the subnetwork using the toolbar](#interact-network)
    1. &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1 - [Adjust network view](#adjust-vew)
    2. &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2 - [Download network](#download)
5. 5 - [Enrichment analysis with ChEA-KG](#chea-enrichment)
	1. &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1 - [Subnetwork](#enrichment-subnetwork)
	2. &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2 - [Search parameters](#enrichment-params)
	3. &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3 - [Perform a query](#perform-query)

<br/><br/>
## **Abstract** <a name="grn-search"></a>
Gene expression is controlled by transcription factors that selectively bind to DNA to regulate mRNA expression of all human genes. Transcription 
factors control the expression of other transcription factors, forming a complex gene regulatory network (GRN) with switches, feedback loops, and 
other regulatory motifs. Many experimental methods and computational tools have been developed to reconstruct GRNs in-silico. Here we present a 
different approach to reconstruct the human GRN. By submitting thousands of gene sets from the RummaGEO resource for transcription factor 
enrichment analysis with ChEA3, we are able to distill signed and directed edges that connect all human transcription factors to construct a high 
quality human GRN. The GRN has 130,793 signed and directed edges between 703 source and 1,543 target transcription factors. The network is made 
accessible via an interactive web-based application called ChEA-KG. ChEA-KG enables users to query the GRN by searching for single or pairs of 
transcription factors, as well as by submitting gene sets to perform transcription factor enrichment analysis with ChEA3 and then place the 
enriched transcription factors in context of ChEA-KG. To demonstrate the utility of ChEA-KG, we systematically identified transcription factor 
subnetworks that regulate differentially expressed genes in tumors from ten cancer types, and 69 subtypes, profiled by the Clinical Proteomic Tumor 
Analysis Consortium (CPTAC).    

## **The ChEA-KG GRN** <a name="introduction"></a>
ChEA-KG visualizes a human GRN that connects **1,554** transcription factors (TFs) with signed, directed edges that indicate up- or down-regulatory 
relationships between TFs. There are **63,375** upregulated edges and **67,418** downregulated edges. Then network is constructed by inputting 171,441 human gene
sets from [RummaGEO](https://rummageo.com) into ChEA-3 for enrichment analysis. For each gene set, edges are identified between each enriched TF, which
serves as the source node, and the TFs in the gene set, which become the targets. The final ChEA-KG network is filtered by edge significaance (p < 0.01).

**Nodes** 

The 1,554 TFs are the TFs found to be highly enriched for any RummaGEO gene set [ChEA3 primary libraries](https://maayanlab.cloud/chea3/). 
Each node is associated with an **id, label**, and **URI** that points to the NCBI gene page for that gene.  


**Edges** 

Edges in the GRN indicate regulatory relationships between source and target TFs, determined using enrichment analysis. There are two edge types (Figure 1):   
    1 - <span style="color: red;">Red</span> plungers indicate downregulation  
    2 - <span style="color: green;">Green</span> arrows indicate upregulation  

<img src="https://minio.dev.maayanlab.cloud/hgrn-chear/tutorial/introduction1.png" alt="drawing" width="400"/> 
**Figure 1.** The ChEA-KG consists of one node type (transcription factors) and two edge types (upregulated and downregulated). 

<p>&nbsp;</p>

## **Searching the KG** <a name="grn-search"></a>
The network search page enables users to query specific subnetworks of the GRN. The search page has three key features: the search panel, the 
network view window, and the toolbar. The search panel provides options to customize the network search, the network view window displays the 
results, and the toolbar provides buttons to interact with the results. The following sections describe how to use these features. 

#### *Single Term Search* <a name="single-search"></a>
To perform a single-term search, start typing in the name of the Entrez gene symbol of a TF into the "Start with" field. If the node exists in the network, it will show up in the autocomplete drop-down menu.
To perform the search, click on the node name in the menu. Alternatively, search either of the two example nodes by clicking on their names. 
This generates a subnetwork of relationships between the search TF and its immediate neighbors.   
<img src="https://minio.dev.maayanlab.cloud/hgrn-chear/tutorial/single-term.png" alt="drawing" width="800"/> 
**Figure 2.** Subnetwork produced by a single-term search for LYL1.  


#### *Expanded Single Term Search* <a name="subgraph-search"></a>
The expanded single term search displays a single-term subnetwork with an additional layer of relationships. To perform this search, input an Entrez
gene symbol into the "Start with" field and click on its name in the autocomplete drop-down menu. Next, toggle the "End Node" switch on (Figure 3). 
Leave the "End with" field **blank**. Control the size of the subnetwork using the slider above the network view. 
<img src="https://minio.dev.maayanlab.cloud/hgrn-chear/tutorial/subgraph.png" alt="drawing" width="800"/> 
**Figure 3.** An expanded LYL1 subnetwork is produced by toggling "End node".   


#### *Two Term Search* <a name="two-term-search"></a>
The two-term search displays shortest path between two transcription factors (Figure 4). In the case where there is a tie for the shortest path 
length, all paths of that length are displayed. To perform this search, enter a starting transcription factor in the "Start with" field. Select its name 
from the autocomplete drop-down menu. Toggle the "End Node" switch, then input a desired end node in the "End with" field. 
<img src="https://minio.dev.maayanlab.cloud/hgrn-chear/tutorial/two-term.png" alt="drawing" width="800"/> 
**Figure 4.** Subnetwork produced by a two-term search for LYL1 and CHCHDC3.  


## **Interacting with the subnetwork using the toolbar** <a name="interact-network"></a>
Basic navigation such as zooming in or out and highlighting nodes and edges are accomplished with the mouse. The toolbar above the network
view provides several buttons for additional methods of interacting with the subnetwork. From left to right, these tools are Size, Full-screen, 
Network View, Table View, Save Subnetwork, Download Graph as an Image File, Show Tooltip, Switch Graph Layout, Show Edge Labels, and Show Legend.     
<p>&nbsp;</p>

#### *Adjust the Network View:* <a name="adjust-view"></a>  
**Basic navigation** 
* To rearrange individual network nodes,* click and drag on any node. Use the mouse to zoom in and out of the network. 
* To emphasize a node and its edges,* hover your mouse over the node. Click on the node to temporarily highlight its edges (Figure 5). 
* Click and drag the whitespace *to move the whole subnetwork.*  
 <img src="https://minio.dev.maayanlab.cloud/hgrn-chear/tutorial/highlight.png" alt="drawing" width="800"/>  
 **Figure 5.** Highlight a node and its edges by clicking on it.  
<p>&nbsp;</p>

**Adjust Subnetwork Size:**  
Adjust the size slider to limit the number of relationships displayed for that network (Figure 6). This adds or subtracts edge relationships from the subnetwork.
 Relationships are prioritized based on their z-score. 
<img src="https://minio.dev.maayanlab.cloud/hgrn-chear/tutorial/network_size.png" alt="drawing" width="800"/> 
**Figure 6.** Adjusting the slider changes the number of relationships displayed subnetwork.  


**Full Screen:**  
Clicking on the full-screen button displays the search panel, network view, and toolbar in full-screen.  
**Enable the Tooltip:**  
The tooltip is used to see more information about a node in the network. The tooltip lists the ID, label, and URI for a node.  view the ID, label, and URI for a node when the mouse hovers over or clicks on a node in the network (Figure 7). 
<img src="https://minio.dev.maayanlab.cloud/hgrn-chear/tutorial/tooltip.png" alt="drawing" width="800"/> 
**Figure 7.** Enabling the tooltip displays the node ID, label, and URI for each node that is clicked or hovered over.  


**Delete or Expand a Node:**  
The tooltip also provides options to add relationships involving that node to the subnetwork results (expand), or to view the single-term search results for that node in a new page (open node in a new page) (Figure 8). 
In the single-term search and expanded single-term search, the tooltip also provides the option to delete nodes from the results (Figure 9).
To use these options, with the tooltip enabled click on the node of interest to view its tooltip. Click on the "delete", "expand", or "open node in a new page" icons to perform the action.  
*Expand the CTCFL node by clicking expand:*  
<img src="https://minio.dev.maayanlab.cloud/hgrn-chear/tutorial/expand1.png" alt="drawing" width="500"/>
<img src="https://minio.dev.maayanlab.cloud/hgrn-chear/tutorial/expand2.png" alt="drawing" width="500"/> 
**Figure 8.** Expand the CTCFL node by clicking expand (top). This displays additional relationships involving the CTCFL node (bottom). 

<p>&nbsp;</p>

*Delete the expanded CTCFL node by clicking delete (Figure 8):*  
<img src="https://minio.dev.maayanlab.cloud/hgrn-chear/tutorial/delete1.png" alt="drawing" width="500"/>
<img src="https://minio.dev.maayanlab.cloud/hgrn-chear/tutorial/delete2.png" alt="drawing" width="500"/>
**Figure 8.** Remove the CTCFL node by clicking delete (top). This removes the CTCFL node and its relationships (bottom).  
  
**Network and Table View:**  
In addition to the network view, there is also the option to view the network as a table. To change between the network and table view, click on their respective buttons (Figure 9). Three tables are produced, one for the nodes and one for each edge type,
which show the data associated with each node and edge. The tables contain the following fields:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- Nodes: Transcription Factor ID, Transcription Factor symbol, URI
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- Edges (upregulated and downregulated):  Source, relation, target, z score, p value
Switch between the table results using the tabs at the top. Invidual entries can be searched by entering an item into the "Search Results" field. 
The table can be exported as a CSV file or printed using the "Export" button to the right of the search bar. Click on any of the header titles to sort the table according to that field. Adjust the 
number of results displayed and view additional entries using the buttons at the bottom of the table. 
<img src="https://minio.dev.maayanlab.cloud/hgrn-chear/tutorial/table_view.png" alt="drawing" width="800"/> 
**Figure 9.** View the network as a table using the Table View button.  


**Switch Graph Layout:**  
The network view provides three graph layouts from Cytoscape. These are force-directed (emphasizes clustering), geometric (organizes nodes into circle; emphasizes density), and hierarchical (useful for tree-like graphs).
More information about graph layout algorithms is available on the [Cytoscape blog](https://blog.js.cytoscape.org/2020/05/11/layouts/#classes-of-layouts). Switch between the graph 
layouts by clicking on the Graph Layout button, and then clicking on the name of the desired layout (Figure 10). 
<img src="https://minio.dev.maayanlab.cloud/hgrn-chear/tutorial/heirarchical-layout.png" alt="drawing" width="800"/> 
**Figure 10.** Switch to a hierarchical layout using the Graph Layout button. 

**Show Edge Labels:**  
To overlay the edge labels ("upregulates" and "downregulates") on the graph edges, click on the Edge Labels button. (Figure 11).
<img src="https://minio.dev.maayanlab.cloud/hgrn-chear/tutorial/edge-label.png" alt="drawing" width="800"/> 
**Figure 11.** Display the edge labels using the Edge Labels button.  

**Show Legend:**  
To display a legend describing the node and edge colors, use the Show Legend button (Figure 12). On clicking this button, an additional toolbar button is displayed that, when clicked, adjusts the size of the legend.  
<img src="https://minio.dev.maayanlab.cloud/hgrn-chear/tutorial/legend.png" alt="drawing" width="800"/> 
**Figure 12.** Show and adjust the size of the legend. 
 

#### *Download the Subnetwork:*  <a name="download"></a>
There are two methods of saving the subnetwork: saving the network to a file or saving the network as an image. 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;*Save subnetwork:* Save the subnetwork to a file. This produces two files:   
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1&nbsp;-&nbsp;nodes.csv has the fields [id, label, kind, uri, color]  
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2&nbsp;-&nbsp;edges.csv has the  fields [source, target, relation, source_label, target_label, kind, p_value, z_score]  
<p>&nbsp;</p>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;*Download graph as an image file:* Clicking on this button shows a menu of image formats. Click on the appropriate one to save an image of the current subnetwork view.  
<p>&nbsp;</p>


## **Enrichment Analysis with ChEA-KG** <a name="chea-enrichment"></a>
The enrichment analysis page provides an option to visualize a subnetwork of the GRN based on ChEA3 enrichment analysis results (Figure 13). 
<img src="https://minio.dev.maayanlab.cloud/hgrn-chear/tutorial/enrichment_home1.png" alt="drawing" width="1200"/> 
**Figure 13.** Switch to a hierarchical layout using the graph layout button.  

### Enrichment Analysis Subnetwork <a name="enrichment-subnetwork"></a>
The subnetwork returned by this feature contains transcription factors with the highest integrated mean rank for that gene set (Figure 14). By default, the top 10 ranked TFs are returned. The ChEA3 integrated mean rank is calculated from the average rank across six libraries: Enrichr Queries, GTEx Coexpression, ARCHS4 Coexpression, ENCODE ChIP-seq, Literature ChIP-seq, and ReMap ChIP-seq. For more information on how the ChEA3 Integrated Mean Rank is calculated, visit the [ChEA3 website](https://maayanlab.cloud/chea3/). 
**Figure 14.** Enrichment analysis produces a subnetwork of enriched TFs for a gene set. 

### Enrichment Analysis Search Parameters: <a name="enrichment-params"></a>
There are three advanced search options that can be adjusted by toggling "Advanced Options".  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1 - Minimum libraries: This filters out nodes from the enrichment results if they occur in fewer than the specified number of ChEA3 libraries (maximum of 6)  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2 - Maximum edge p-value: Specify the maximum p-value of an edge.  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3 - Specify number of nodes: Specify how many top ranked nodes.  

### Perform a Query:<a name="perform-query"></a>
To perform a query, input a list of newline-separated Entrez gene symbols into the text box. Alternatively, click "try an example" to use an example gene set. Add a description under the "Description" field. Adjust the advanced options as necessary. Click submit. 